cpp/models/ode_secir/parameters.h

```diff
@@ -360,201 +360,236 @@ class Parameters : public ParametersBase
      */
-    void apply_constraints()
+    bool apply_constraints()
     {
+        double tol_times = 1e-4; // accepted tolerance for compartment stays
+
+        int corrected = false;
         if (this->get<Seasonality>() < 0.0 || this->get<Seasonality>() > 0.5) {
             log_warning("Constraint check: Parameter Seasonality changed from {:0.4f} to {:d}",
                         this->get<Seasonality>(), 0);
             this->set<Seasonality>(0);
+            corrected = true;
         }
 
         if (this->get<ICUCapacity>() < 0.0) {
             log_warning("Constraint check: Parameter ICUCapacity changed from {:0.4f} to {:d}",
                         this->get<ICUCapacity>(), 0);
             this->set<ICUCapacity>(0);
+            corrected = true;
         }
 
         for (auto i = AgeGroup(0); i < AgeGroup(m_num_groups); ++i) {
 
-            if (this->get<IncubationTime>()[i] < 2.0) {
+            if (this->get<IncubationTime>()[i] < tol_times) {
                 log_warning("Constraint check: Parameter IncubationTime changed from {:.4f} to {:.4f}",
-                            this->get<IncubationTime>()[i], 2.0);
-                this->get<IncubationTime>()[i] = 2.0;
+                            this->get<IncubationTime>()[i], tol_times);
+                this->get<IncubationTime>()[i] = tol_times;
+                corrected                      = true;
             }
 
-            if (2 * this->get<SerialInterval>()[i] < this->get<IncubationTime>()[i] + 1.0) {
+            if (2 * this->get<SerialInterval>()[i] < this->get<IncubationTime>()[i] + tol_times) {
                 log_warning("Constraint check: Parameter SerialInterval changed from {:.4f} to {:.4f}",
-                            this->get<SerialInterval>()[i], 0.5 * this->get<IncubationTime>()[i] + 0.5);
-                this->get<SerialInterval>()[i] = 0.5 * this->get<IncubationTime>()[i] + 0.5;
+                            this->get<SerialInterval>()[i], 0.5 * this->get<IncubationTime>()[i] + tol_times / 2);
+                this->get<SerialInterval>()[i] = 0.5 * this->get<IncubationTime>()[i] + tol_times / 2;
+                corrected                      = true;
             }
-            else if (this->get<SerialInterval>()[i] > this->get<IncubationTime>()[i] - 0.5) {
+            else if (this->get<SerialInterval>()[i] > this->get<IncubationTime>()[i] - tol_times / 2) {
                 log_warning("Constraint check: Parameter SerialInterval changed from {:.4f} to {:.4f}",
-                            this->get<SerialInterval>()[i], this->get<IncubationTime>()[i] - 0.5);
-                this->get<SerialInterval>()[i] = this->get<IncubationTime>()[i] - 0.5;
+                            this->get<SerialInterval>()[i], this->get<IncubationTime>()[i] - tol_times / 2);
+                this->get<SerialInterval>()[i] = this->get<IncubationTime>()[i] - tol_times / 2;
+                corrected                      = true;
```

Comment:
Does this combination of SI and IP not lead to a division by zero if IP=1e-4? Can you check the relation on SI and IP?

Reply:
[constraints_SI_INC.pdf](https://github.com/DLR-SC/memilio/files/12248347/constraints_SI_INC.pdf)


Here is a simple (not really nice) sheet on which I have shown the derivation of the constraints